File: Day110_2025/Q110_Leetcode.py
from typing import List
from math import ceil
from collections import Counter

class Solution:
    def numRabbits(self, answers: List[int]) -> int:
        freq = Counter(answers)
        total = 0

        for x, count in freq.items():
            group_size = x + 1
            groups = ceil(count / group_size)
            total += groups * group_size

        return total

File: Day110_2025/Q110_Leetcode.java
class Q110_Leetcode {
    public int numRabbits(int[] answers) {
        Map<Integer, Integer> freqMap = new HashMap<>();
        for (int ans : answers) {
            freqMap.put(ans, freqMap.getOrDefault(ans, 0) + 1);
        }

        int total = 0;
        for (Map.Entry<Integer, Integer> entry : freqMap.entrySet()) {
            int x = entry.getKey();      // Rabbit says "x others like me"
            int count = entry.getValue(); // Number of such replies
            int groupSize = x + 1;
            int groups = (int) Math.ceil((double) count / groupSize);
            total += groups * groupSize;
        }
        return total;
    }
}
